include/common/result.hpp

```diff
@@ -0,0 +1,146 @@
+        : value(std::forward<U>(value)) {}
+
+    // Construct a Result with a value and an error value.
+    // Constraint: the value member variable must be able to be constructed with the value
+    // parameter.
+    // Constraint: the error parameter must be of a type that may be contained in the error member
+    // variable.
+    template<typename U, CustomError E>
+        requires std::constructible_from<T, U>
+                     && (std::is_same_v<Errs, std::remove_cvref<E>> || ...)
+    constexpr Result(U&& value, E&& error)
+        : value(std::forward<U>(value)),
+          error(std::forward<E>(error)) {}
+
+    // Construct a Result with an error, initializing the normal value to its sentinel value.
+    // Constraint: the type of the normal value must have a specialized sentinel value.
+    // Constraint: the error parameter must be of a type that may be contained in the error member
+    // variable.
+    template<CustomError U>
+        requires Sentinel<T> && (std::is_same_v<Errs, std::remove_cvref<U>> || ...)
+    constexpr Result(U&& error)
+        : value(sentinel_v<T>),
+          error(std::forward<U>(error)) {}
+
+    // Check whether the result contains an error of the given type
+    // Constraint: the custom error type to check for must be able to be contained in the error
+    // member variable.
+    template<CustomError U>
+        requires(std::is_same_v<Errs, std::remove_cvref<U>> || ...)
+    constexpr U has() const& {
+        if (!error.has_value()) {
+            return false;
+        } else {
+            return std::holds_alternative<U>(error);
+        }
+    };
+
+    // return an optional wrapping the given error type.
+    // Constraint: the custom error type to get must be able to be contained in the error
+    // member variable.
+    template<CustomError U>
+        requires(std::is_same_v<Errs, std::remove_cvref<U>> || ...)
+    constexpr std::optional<U> get() const& {
+        if (this->has<U>()) {
+            return std::get<U>(error.value());
+        } else {
+            return std::nullopt;
+        }
+    }
+
+    // return an optional wrapping the given error type.
+    // Constraint: the custom error type to get must be able to be contained in the error
+    // member variable.
+    template<CustomError U>
+        requires(std::is_same_v<Errs, std::remove_cvref<U>> || ...)
+    constexpr std::optional<U> get() && {
+        if (this->has<U>()) {
+            return std::get<U>(std::move(error.value()));
+        } else {
+            return std::nullopt;
+        }
+    }
+
+    // implicit conversion operator, on an l-value
+    constexpr operator T() const& {
+        return value;
+    };
+
+    // implicit conversion operator, on an r-value
+    constexpr operator T() && {
+        return std::move(value);
+    }
+
+    // the optional error value.
+    // a variant that could contain any of the specified error types.
+    std::optional<std::variant<Errs...>> error;
```

Comment:
I think its more space efficient for this to be `std::variant<std::monotype, Errs...>`, thats sort of what std::monotype is for. 

(unless stdlib already has a partial spec of `std::optional<std::variant<Ts...>>` optimizing for this... but i dont think so).